src/Simplex/Chat.hs

```diff
@@ -5524,46 +5525,67 @@ directMessage chatMsgEvent = do
-deliverMessage conn@Connection {connId} cmEventTag msgBody msgId = do
-  let msgFlags = MsgFlags {notification = hasNotification cmEventTag}
-  agentMsgId <- withAgent $ \a -> sendMessage a (aConnId conn) msgFlags msgBody
-  let sndMsgDelivery = SndMsgDelivery {connId, agentMsgId}
-  withStore' $ \db -> createSndMsgDelivery db sndMsgDelivery msgId
+deliverMessage conn cmEventTag msgBody msgId =
+  deliverMessages [(conn, cmEventTag, msgBody, msgId)] >>= \case
+    [Right msgDeliveryId] -> pure msgDeliveryId
+    [Left e] -> throwError e
+    rs -> throwChatError $ CEInternalError $ "deliverMessage: expected 1 result, got " <> (show $ length rs)
+
+deliverMessages :: ChatMonad' m => [(Connection, CMEventTag e, MsgBody, MessageId)] -> m [Either ChatError Int64]
+deliverMessages [] = pure []
+deliverMessages msgReqs = do
+  rs <- replicateM (length msgReqs) (newIORef $ Left $ ChatError $ CEInternalError "skipped in batch")
+  let aReqs = map (\(conn, cmEvTag, msgBody, _) -> (aConnId conn, msgFlags cmEvTag, msgBody)) msgReqs
+  rs' <- withAgent' (`sendMessages` aReqs)
+  rs1 <- catMaybes <$> mapM processResults (zip3 rs msgReqs rs')
+  rs1' <- withStoreBatch' (\db -> map (createDelivery db) rs1)
+  mapM_ (\((r, _, _), msgDeliveryId) -> writeIORef r msgDeliveryId) (zip rs1 rs1')
+  mapM readIORef rs
+  where
+    msgFlags cmEvTag = MsgFlags {notification = hasNotification cmEvTag}
+    processResults (r, req, aMsgId_) = case aMsgId_ of
+      Left e -> Nothing <$ writeIORef r (Left $ ChatErrorAgent e Nothing)
+      Right aMsgId -> pure $ Just (r, req, aMsgId)
+    createDelivery db (_, (Connection {connId}, _, _, msgId), agentMsgId) =
+      createSndMsgDelivery db (SndMsgDelivery {connId, agentMsgId}) msgId
 
 sendGroupMessage :: (MsgEncodingI e, ChatMonad m) => User -> GroupInfo -> [GroupMember] -> ChatMsgEvent e -> m (SndMessage, [GroupMember])
-sendGroupMessage user GroupInfo {groupId} members chatMsgEvent =
-  sendGroupMessage' user members chatMsgEvent groupId Nothing $ pure ()
-
-sendGroupMessage' :: forall e m. (MsgEncodingI e, ChatMonad m) => User -> [GroupMember] -> ChatMsgEvent e -> Int64 -> Maybe Int64 -> m () -> m (SndMessage, [GroupMember])
-sendGroupMessage' user members chatMsgEvent groupId introId_ postDeliver = do
-  msg <- createSndMessage chatMsgEvent (GroupId groupId)
-  -- TODO collect failed deliveries into a single error
+sendGroupMessage user GroupInfo {groupId} members chatMsgEvent = do
+  msg@SndMessage {msgId, msgBody} <- createSndMessage chatMsgEvent (GroupId groupId)
   recipientMembers <- liftIO $ shuffleMembers (filter memberCurrent members) $ \GroupMember {memberRole} -> memberRole
-  rs <- forM recipientMembers $ \m ->
-    messageMember m msg `catchChatError` (\e -> toView (CRChatError (Just user) e) $> Nothing)
-  let sentToMembers = catMaybes rs
+  let ms = mapMaybe (memberToSend chatMsgEvent members) recipientMembers
+      tag = toCMEventTag chatMsgEvent
+      (pending, toSend) = foldl' addMember ([], []) ms
```

Comment:
this seems to somewhat repeat memberToSend logic, why not accumulate and return these lists on the first members pass?